enhancements/scheduling/clusteroperator-scheduling.md

```diff
@@ -0,0 +1,108 @@
+---
+title: clusteroperator-scheduling
+authors:
+  - "@michaelgugino"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-02-12
+last-updated: 2021-02-12
+status: implementable
+---
+
+# ClusterOperator Scheduling
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+To ensure system components, especially ClusterOperators, have the correct
+set of tolerations.
+
+## Motivation
+
+The motivation is to create an easy to follow reference for developing and
+updating ClusterOperators and similar system components.
+
+### Goals
+
+* List best-practices for tolerations
+
+### Non-Goals
+
+* Non-standard cluster profiles such as single node
+* Add all practices here today, we can add them over time.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want upgrades to proceed safely and without
+intervention.
+
+#### Story 2
+
+As an OpenShift developer, I want to be able to reference a guide to ensuring
+my ClusterOperators are properly defined for scheduling purposes.
+
+### Implementation Details/Notes/Constraints [optional]
+
+Not all elements of scheduling are defined here.  There are several, and we
+can update this document as needed.
+
+### Risks and Mitigations
+
+Not all ClusterOperators behave identically.  Careful consideration for each
+ClusterOperator is paramount.
+
+## Design Details
+
+### NoSchedule Tolerations
+
+#### node.kubernetes.io/unschedulable
+
+There are various `NoSchedule` taints that are commonly applied to nodes.
+ClusterOperators should not tolerate NoSchedule taints with key:
+`node.kubernetes.io/unschedulable`
+
+This is a special key that is applied when a node is cordoned, such as during
+drain.  Since this is a NoSchedule taint, anything without this taint will
+not automatically be evicted, it will only prevent future scheduling.
+
+During upgrades, we cordon and drain all nodes.  Sometimes the scheduler will
+place pods tolerating NoSchedule onto a node quicker than the loop for
+draining a node can complete.  This results in hot-looping between draining
+and the scheduler.
+
+Any other NoSchedule toleration is fine.
```

Comment:
In the spirit of Chesterton fence, I’d like to see a review of the components that thought they *should* set this and why (representative sample).

That ensures we correct misconceptions about when the toleration is appropriate.

If it was accidental, what would have prevented the accident?

Reply:
The one I know of for sure is the CVO.  I don't have data if anything else is doing it or not.  I will take a look.

Reply:
Looks like the CVO is the only-non daemonset pod that has this toleration, at least from a typical CI run.  The daemonsets get that toleration dynamically, so everything else looks correct.

Reply:
There was a discussion [sometime ago](https://github.com/openshift/cluster-version-operator/pull/182#discussion_r280956544) on whether CVO should tolerate `NoSchedule` taint and the answer was `yes`. Has something changed now? I was always in favour of specific tolerations instead of blanket tolerations. How about adding a taint called `drain-in-progress` to node with `NoSchedule` effect. That won't be tolerated by CVO which should help in proceeding with draining smoothly.

Reply:
I am not entire sure of that, IIUC I got a firm `yes` on `unschedulable` however if you think this was not needed, I can change the tolerations of CVO.

Reply:
There are many taints that have the effect 'NoSchedule' but there is one 'node.kubernetes.io/unschedulable' taint with the effect 'NoSchedule'.  It appears in the previous discussion only NoSchedule was considered rather than the specific taints (at the least, node.kubernetes.io/unschedulable was not itself considered directly).

Reply:
any `NoSchedule` taint != `unschedulable` condition. `unschedulable` condition == `node.kubernetes.io/unschedulable` taint. unschedulable is a node condition which is being represented via a special well known taint called `node.kubernetes.io/unschedulable`.

Reply:
I'm trying to use the k8s precise language.  There are taints, and those taints have effects.  The effect we're discussing is NoSchedule.  That effect is caused by one or more taints.

I did not see the taint node.kubernetes.io/unschedulable discussed in the link you provided.  There were only discussions around the NoSchedule effect.